File: hibernate-criteria-builder/src/main/java/io/github/robertomike/hefesto/constructors/ConstructGroupByImplementation.java
package io.github.robertomike.hefesto.constructors;

import lombok.Getter;

import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Expression;
import javax.persistence.criteria.Root;

@Getter
public class ConstructGroupByImplementation extends ConstructGroupBy {
    public void construct(CriteriaQuery<?> cr, Root<?> root) {
        if (items.isEmpty()) {
            return;
        }

        cr.groupBy(items.stream().map(value -> root.get(value.field())).toArray(Expression[]::new));
    }
}

File: hibernate-criteria-builder/src/main/kotlin/io/github/robertomike/hefesto/constructors/ConstructGroupByImplementation.kt
package io.github.robertomike.hefesto.constructors

import jakarta.persistence.criteria.CriteriaQuery
import jakarta.persistence.criteria.Expression
import jakarta.persistence.criteria.Root

/**
 * Criteria API implementation of GROUP BY clause construction.
 * 
 * Converts GROUP BY definitions into JPA Criteria API groupBy expressions.
 * Groups query results by specified fields, typically used with aggregate functions.
 * 
 * Example:
 * ```kotlin
 * query
 *     .select("status", "count")
 *     .selectCount("id")
 *     .groupBy("status")
 * ```
 */
class ConstructGroupByImplementation : ConstructGroupBy() {
    /**
     * Constructs and applies GROUP BY clauses to the CriteriaQuery.
     * Maps field names to root entity expressions and groups results.
     *
     * @param cr the CriteriaQuery to apply grouping to
     * @param root the root entity
     */
    fun construct(cr: CriteriaQuery<*>, root: Root<*>) {
        if (items.isEmpty()) {
            return
        }

        cr.groupBy(*items.map { value -> root.get<Any>(value.field) }.toTypedArray())
    }
}
